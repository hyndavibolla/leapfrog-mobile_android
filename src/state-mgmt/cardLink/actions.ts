import { CardLink } from '_models';

export enum ACTION_TYPE {
  SET_LOCAL_OFFERS = '[cardlink] set local offers',
  SET_LINKED_CARDS = '[core] set linked card list',
  SET_LAST_SEARCHED_LOCATION = '[cardlink] set last searched location',
  ACTIVATE_LOCAL_OFFER = '[cardlink] activate local offer',
  SET_LOCAL_OFFER_FAILED = '[cardlink] set isLocalOfferFailed',
  SET_HAS_CALLED_LOCAL_OFFER = '[cardlink] set hasCalledLocalOffer',
  SET_ROUTE_TO_ACTIVATE_LOCAL_OFFER = '[cardlink] set routeToActivateLocalOffer'
}

export const actions = {
  setLocalOffers: (localOffers: CardLink.ICardLinkOffers) => ({ type: ACTION_TYPE.SET_LOCAL_OFFERS, payload: localOffers }),
  setLinkedCards: (linkedCardsList: CardLink.ILinkedCard[]) => ({ type: ACTION_TYPE.SET_LINKED_CARDS, payload: { linkedCardsList } }),
  setLastSearchedLocation: (location: CardLink.ILocation) => ({ type: ACTION_TYPE.SET_LAST_SEARCHED_LOCATION, payload: { location } }),
  activateLocalOffer: (offerId: string) => ({ type: ACTION_TYPE.ACTIVATE_LOCAL_OFFER, payload: offerId }),
  setLocalOfferFailed: (isLocalOfferFailed: boolean) => ({ type: ACTION_TYPE.SET_LOCAL_OFFER_FAILED, payload: isLocalOfferFailed }),
  setHasCalledLocalOffer: (hasCalledLocalOffer: boolean) => ({ type: ACTION_TYPE.SET_HAS_CALLED_LOCAL_OFFER, payload: hasCalledLocalOffer }),
  setRouteToActivateLocalOffer: (routeToActivateLocalOffer: string) => ({
    type: ACTION_TYPE.SET_ROUTE_TO_ACTIVATE_LOCAL_OFFER,
    payload: routeToActivateLocalOffer
  })
};
