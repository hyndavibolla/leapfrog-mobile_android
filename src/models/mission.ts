import { HTMLString, DateLike } from './general';
import { ProgramType } from './offer';

export interface IMission {
  uuid: string; // generated on the client (api service) by combining offerCode with offerId
  name: string;
  description: string;
  image?: string;
  termsAndConditions: HTMLString;
  brandName: string;
  brandLogo?: string;
  brandDescription: string;
  brandCategories: string[];
  brandRequestorId: string;
  offerId: string;
  offerCode: string;
  rank: number;
  validFrom: DateLike;
  validUntil: DateLike;
  pointsAwarded: IPointsAwarded;
  callToActionUrl: string;
  provider: Provider;
  pubRef: string;
  rewardOfferCode: string;
}

export enum Provider {
  BUTTON = 'button',
  SURVEY = 'AdBloom',
  CARDLINK = 'mastercard'
}

export enum RedemptionType {
  POINT_PER_DOLLAR = 'PointsPerDollar',
  FIXED_POINTS = 'FixedPoints',
  PERCENT_OFF = 'PercentOff'
}

export interface IPointsAwarded {
  rewardType: RedemptionType;
  rewardValue: number;
  conditions: IAwardCondition[];
  alternateRewardValue?: number;
  programType?: ProgramType;
  programSubType?: string;
}

export interface IAwardCondition {
  category?: string;
  rewardValue: number;
  rewardType: RedemptionType;
  rewardOfferCode?: string;
  programType?: ProgramType;
  programSubType?: string;
  alternateRewardValue?: number;
}

export enum KeywordType {
  CATEGORY = 'category',
  BRAND = 'brand'
}

export interface IKeyword {
  value: string;
  type: KeywordType;
}

export interface ICategory {
  name: string;
  imageUrl: string;
  lifestyleUrl: string;
  rank: number;
}

export enum MissionListType {
  DEFAULT = 'Default',
  KEEP_EARNING = 'KeepEarning',
  FEATURED = 'Featured',
  SPECIAL = 'Specials',
  JUST_FOR_YOU = 'JustForYou',
  TRENDING = 'Trending',
  DYNAMIC_LIST_2 = 'EarnList2',
  DYNAMIC_LIST_3 = 'EarnList3',
  DYNAMIC_LIST_4 = 'EarnList4',
  DYNAMIC_LIST_6 = 'EarnList6',
  EXCEPTIONAL = 'EarnList5'
}

export enum RecentSearchHistoryType {
  MISSION = 'mission',
  CATEGORY = 'category '
}
export interface IRecentSearchHistory {
  id: string;
  name: string;
  type: RecentSearchHistoryType;
  isAvailableStreakIndicator?: boolean;
}
