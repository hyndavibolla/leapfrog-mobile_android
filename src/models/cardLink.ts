import { DateLike } from './general';
import { RedemptionType } from './mission';

export enum DayOfWeek {
  SUNDAY = 'Sun',
  MONDAY = 'Mon',
  TUESDAY = 'Tue',
  WEDNESDAY = 'Wed',
  THURSDAY = 'Thu',
  FRIDAY = 'Fri',
  SATURDAY = 'Sat'
}

export interface ICardLinkOfferCalendarDayHours {
  open: DateLike;
  close: DateLike;
  openForBusiness: boolean;
}

export interface ICardLinkOfferCalendarBenefits {
  available: boolean;
  monthDay: Date;
}

export interface ICardLinkOfferCalendar {
  dayOfWeek: DayOfWeek;
  dayHours: ICardLinkOfferCalendarDayHours;
}

export interface ICardLinkOfferMerchant {
  address: {
    latitude: number;
    longitude: number;
    street: string;
  };
  priceRange: string;
  merchantDistance: number;
  websiteUrl: string;
}

export interface ICardLinkOfferPointsAwarded {
  rewardValue: number;
  rewardType: RedemptionType;
}

export enum ProgramType {
  SYWMAX = 'SYWMAX'
}

export enum ProgramSubType {
  MC = 'MC'
}

export enum Providers {
  MASTERCARD = 'mastercard'
}

export enum InStoreOfferStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive'
}

export interface ICardLinkOffer {
  activeUntil: DateLike;
  benefits: ICardLinkOfferCalendarBenefits[];
  brandImage: string;
  brandName: string;
  brandLogo: string;
  calendar: ICardLinkOfferCalendar[];
  /** @deprecated For local offers, use merchant.websiteUrl instead. */
  callToActionUrl: string;
  description: string;
  disclaimer: string;
  menu: {
    externalUrl: string;
  };
  merchant: ICardLinkOfferMerchant;
  offerId: string;
  pointsAwarded: ICardLinkOfferPointsAwarded;
  programType: ProgramType;
  programSubType: ProgramSubType;
  provider: Providers;
  rating: {
    overallRating: number;
  };
  status: InStoreOfferStatus;
  validFrom: DateLike;
  validUntil: DateLike;
}

export interface ICardLinkOffers {
  userId: string;
  offers: ICardLinkOffer[];
}

export interface ILinkedCard {
  cardId: string;
  cardLastFour: string;
  cardType: string;
  partnerType: LinkedCardPartnerType;
  tuId?: string;
  isSywCard: boolean;
}

export enum LinkedCardPartnerType {
  MASTERCARD = 'MSTR',
  REWARDS_NETWORK = 'RWNW'
}

export enum EnrollmentEventType {
  ENROLL = 'ENROLL',
  UNENROLL = 'UNENROLL'
}

export enum cardName {
  VISA = 'Visa',
  MSTR = 'MasterCard',
  DISC = 'Discovery',
  AMEX = 'American Express'
}

export const PRICE_RANGE_MAX_VALUE = 5;

export interface IGeocodeAddress {
  formatted_address: string;
  geometry: IGeometry;
  types: string[];
  address_components: IAddressComponent[];
  parsedAddress?: IParsedAddress;
}

export interface IParsedAddress {
  postalCode?: IAddressComponent;
  city?: IAddressComponent;
  state?: IAddressComponent;
  country?: IAddressComponent;
}
export interface IGeometry {
  location: IGeocodeLocation;
}

export interface IGeocodeLocation {
  lat: number;
  lng: number;
}

export interface ILocation {
  latitude: number;
  longitude: number;
}

export interface IAddressComponent {
  long_name: string;
  short_name: string;
  types: string[];
}

export enum LocationType {
  ADDRESS = 'Address',
  OFFER = 'Offer'
}
export interface ILocationStorage {
  type: LocationType;
  value: string;
  searchItem: IGeocodePlacePrediction | string;
  timestamp: number;
}
export interface IGeocodePlace {
  predictions: IGeocodePlacePrediction[];
}

export interface IGeocodePlacePrediction {
  place_id: string;
  structured_formatting: IGeocodePlaceStructuredFormatting;
  types: string[];
  place_details?: IGeocodePlaceDetails;
}

export interface IGeocodePlaceStructuredFormatting {
  main_text: string;
  secondary_text: string;
  main_text_matched_substrings: IGeocodePlaceStructuredFormattingMatchedSubstrings[];
}

export interface IGeocodePlaceStructuredFormattingMatchedSubstrings {
  offset: number;
  length: number;
}

export interface IGeocodePlaceDetails {
  formatted_address: string;
  geometry: IGeometry;
  place_id: string;
  name: string;
}
