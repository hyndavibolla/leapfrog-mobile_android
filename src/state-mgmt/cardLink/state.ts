import { ICardLinkOffer, ILinkedCard, ILocation } from '_models/cardLink';

export interface ICardLinkState {
  userId: string;
  offers: ICardLinkOffer[];
  linkedCardsList: ILinkedCard[];
  lastSearchedLocation: ILocation;
  isLocalOfferFailed: boolean;
  hasCalledLocalOffer: boolean; // to check if at least one time has been called getLocalOffers
  routeToActivateLocalOffer: string; // to activate the local offer after activated a new credit card
}

export const initialState: ICardLinkState = {
  userId: null,
  offers: [],
  linkedCardsList: [],
  lastSearchedLocation: null,
  isLocalOfferFailed: false,
  hasCalledLocalOffer: false,
  routeToActivateLocalOffer: ''
};
