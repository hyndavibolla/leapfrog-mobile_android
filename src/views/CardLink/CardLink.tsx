import React, { memo, useContext, useMemo, useState, useCallback, useEffect } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { WebViewNavigation } from 'react-native-webview';

import { CriticalError } from '_components/CriticalError';
import { WebView } from '_components/WebView/WebView';
import { ConnectionBanner } from '_components/ConnectionBanner';

import { GlobalContext } from '_state_mgmt/GlobalState';
import { KnownMissionSearchKey } from '_state_mgmt/mission/state';
import { useGetLinkedCardsList } from '_state_mgmt/cardLink/hooks';
import { useTestingHelper } from '_utils/useTestingHelper';
import { Provider } from '_models/mission';
import { ENV } from '_constants';
import { actions } from '_state_mgmt/cardLink/actions';

export interface Props {
  navigation: StackNavigationProp<any>;
  route: { params?: { shouldActivateOffer?: boolean } };
}

const { MAX_LOAD_ATTEMPTS: maxLoadAttempts, ATTEMPT_DELAY_MS: attemptDelay } = ENV.WEBVIEWS;
const { ENROLLMENT_FINISH_URI_REGEX: enrollmentFinishUriRegex } = ENV.LOCAL_OFFERS;

export const CardLink = ({ navigation, route }: Props) => {
  const { deps, dispatch } = useContext(GlobalContext);
  const [isHttpError, setIsHttpError] = useState(false);
  const [onLoadLinkedCardsList] = useGetLinkedCardsList();
  const [navigationLoaded, setNavigationLoaded] = useState<boolean>(false);
  const shouldActivateOffer = route.params?.shouldActivateOffer ?? false;

  const {
    state: {
      core: { routeHistory },
      mission: { missionSearchMap, missionMap }
    }
  } = useContext(GlobalContext);

  const { getTestIdProps } = useTestingHelper('cardlink');

  const cardlinkOffer = useMemo(
    () =>
      missionSearchMap[KnownMissionSearchKey.EXCEPTIONAL].map(key => missionMap[key]).find(m => m?.provider.toLowerCase() === Provider.CARDLINK.toLowerCase()),
    [missionMap, missionSearchMap]
  );

  useEffect(() => {
    if (!cardlinkOffer?.callToActionUrl) {
      navigation.goBack();
    }
  }, [cardlinkOffer?.callToActionUrl, navigation]);

  const onNavigationStateChange = useCallback(
    async (event: WebViewNavigation) => {
      deps.logger.debug('onNavigationStateChange', { url: event.url });

      const matchesReturnRegex = new RegExp(enrollmentFinishUriRegex).test(event.url);
      deps.logger.debug(`URL ${matchesReturnRegex ? 'matches' : 'does not match'} a returning URL from CardLink.`);
      if (matchesReturnRegex && !navigationLoaded) {
        setNavigationLoaded(true);
        await onLoadLinkedCardsList();
        if (shouldActivateOffer) dispatch(actions.setRouteToActivateLocalOffer(routeHistory[1]));
        navigation.goBack();
      }
    },
    [deps.logger, dispatch, navigation, navigationLoaded, onLoadLinkedCardsList, routeHistory, shouldActivateOffer]
  );

  return (
    <>
      <ConnectionBanner />
      {!cardlinkOffer || isHttpError ? (
        <CriticalError />
      ) : (
        <WebView
          {...getTestIdProps('webview')}
          shouldShowLoader
          source={{ uri: cardlinkOffer.callToActionUrl }}
          retry={{ attempts: maxLoadAttempts, delayMs: attemptDelay }}
          onHttpError={() => {
            setIsHttpError(true);
          }}
          onNavigationStateChangeFN={onNavigationStateChange}
        />
      )}
    </>
  );
};

export default memo(CardLink);
