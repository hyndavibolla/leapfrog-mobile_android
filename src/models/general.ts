import { Dispatch } from 'react';
import Auth0 from 'react-native-auth0';

import { ApiService } from '_services/ApiService';
import { AppService } from '_services/AppService';
import { CCPAService } from '_services/CCPAService';
import { ICrashlyticsService } from '_services/CrashlyticsService';
import { EventTrackerService } from '_services/EventTrackerService';
import { IHttpService } from '_services/HttpService';
import { NativeHelperService } from '_services/NativeHelperService';
import { Logger } from '_services/Logger';
import { IPerformanceTrackingService } from '_services/PerformanceTrackingService';
import { RaiseService } from '_services/RaiseService';
import { RemoteConfigService } from '_services/RemoteConfigService';
import { ICardLinkState } from '_state_mgmt/cardLink/state';
import { ICoreState } from '_state_mgmt/core/state';
import { IGameState } from '_state_mgmt/game/state';
import { IGiftCardState } from '_state_mgmt/giftCard/state';
import { IMessagesState } from '_state_mgmt/messages/state';
import { IMissionState } from '_state_mgmt/mission/state';
import { IRewardState } from '_state_mgmt/reward/state';
import { ISurveyState } from '_state_mgmt/survey/state';
import { IUserState } from '_state_mgmt/user/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IReducer<S = any, A = any> = (state: S, action: IAction<A>) => S;

export interface Deps {
  apiService: ApiService;
  appService: AppService;
  crashlyticsService: ICrashlyticsService;
  httpService: IHttpService;
  performanceTrackingService: IPerformanceTrackingService;
  raiseService: RaiseService;
  nativeHelperService: NativeHelperService;
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
  logger: Logger;
  eventTrackerService: EventTrackerService;
  ccpaService: CCPAService;
  remoteConfigService: RemoteConfigService;
  authService: Auth0;
}
export type IGlobalState = {
  core: ICoreState;
  user: IUserState;
  game: IGameState;
  mission: IMissionState;
  reward: IRewardState;
  survey: ISurveyState;
  messages: IMessagesState;
  cardLink: ICardLinkState;
  giftCard: IGiftCardState;
};

export interface IGlobalContext {
  state: IGlobalState;
  dispatch: Dispatch<IAction<any, any>>;
  deps: Deps;
}

export enum HTTP_METHOD {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE'
}

export interface ITokenSet {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiryTime: string;
  sywToken?: string;
  /**
   * @deprecated deviceId does not belong to ITokenSet anymore.
   */
  deviceId?: string;
  scope?: string;
  expiresIn?: number;
  tokenType?: string;
}

export enum FeatureFlag {
  API_OVERRIDE = 'API_OVERRIDE',
  TEST_ONLY = 'TEST_ONLY',
  ENV_BANNER = 'ENV_BANNER',
  BUTTON_DEBUG = 'BUTTON_DEBUG',
  LOG_LEVEL_INFO = 'LOG_LEVEL_INFO',
  LOG_LEVEL_DEBUG = 'LOG_LEVEL_DEBUG',
  LOG_LEVEL_WARN = 'LOG_LEVEL_WARN',
  LOG_LEVEL_ERROR = 'LOG_LEVEL_ERROR',
  LOG_VIEW = 'LOG_VIEW',
  STREAK_REDEEM_CARD = 'STREAK_REDEEM_CARD',
  SAILTHRU_TRACKING = 'SAILTHRU_TRACKING',
  SURVEY = 'SURVEY',
  PREDICTIVE_MAP_SEARCH = 'PREDICTIVE_MAP_SEARCH',
  MISSIONS_SEARCH = 'MISSIONS_SEARCH',
  SEE_ALL_TRANSACTIONS = 'SEE_ALL_TRANSACTIONS',
  WALLET_YOUR_GC = 'WALLET_YOUR_GC'
}

export type DateLike = Date | string | number;

export type HTMLString = string;

export interface IGameLevelUI {
  updatedAt: DateLike;
  level: number;
}

export interface ICCPASetting {
  allow: boolean;
}

export interface IOnboarding {
  seenOnboarding: boolean;
}

export enum ButtonCreativeType {
  HERO = 'hero',
  CAROUSEL = 'carousel',
  LIST = 'list',
  GRID = 'grid',
  DETAIL = 'detail',
  OTHER = 'other'
}

export interface ISearchHistoryItem {
  keyword: string;
  lastUpdatedAt: DateLike;
}

export enum TooltipKey {
  EARN = 'earn',
  REWARDS = 'rewards',
  MISSIONS = 'missions',
  STREAK = 'streak'
}

export interface IApiOverrideRule {
  key: string;
  description?: string;
  matcher: string; // reg exp
  status: number;
  response: any;
}

export namespace RemoteConfigModel {
  export interface IAppVersion {
    suggested: number;
    required: number;
  }

  export interface IEarnSection {
    name: string;
    order: number;
    visible: boolean;
    title: string;
    description: string;
    seeAllButton: boolean;
  }
}

export enum RUNTIME_ENV {
  LIVE = 'live',
  DEV_TEST = 'test'
}

export interface IMeasure {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface IExpirePointsBannerDataSet {
  lastPointsDate: DateLike;
  lastAcceptedDate: DateLike;
}

export interface ISailthruMessage {
  createdAt: string;
  custom: {
    [k: string]: string;
  };
  htmlText: string;
  id: string;
  isRead: boolean;
  text: string;
  title: string;
  type: string;
  cardImageUrl?: string;
  share: boolean;
}

export interface ITutorialStatus {
  isBannerWatched: boolean;
  date: string;
}
