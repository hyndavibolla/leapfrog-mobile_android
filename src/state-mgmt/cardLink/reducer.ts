import moment from 'moment';
import { ENV } from '_constants';
import { InStoreOfferStatus } from '_models/cardLink';
import { IAction } from '_models/general';
import { ACTION_TYPE } from './actions';
import { ICardLinkState, initialState } from './state';

export const reducer = (state: ICardLinkState = initialState, { type, payload }: IAction): ICardLinkState => {
  switch (type) {
    case ACTION_TYPE.SET_LOCAL_OFFERS:
      return { ...state, userId: payload.userId, offers: payload.offers };
    case ACTION_TYPE.SET_LINKED_CARDS:
      return { ...state, linkedCardsList: payload.linkedCardsList };
    case ACTION_TYPE.SET_LAST_SEARCHED_LOCATION:
      return { ...state, lastSearchedLocation: payload.location };
    case ACTION_TYPE.ACTIVATE_LOCAL_OFFER:
      return {
        ...state,
        offers: state.offers.map(offer =>
          offer.offerId === payload
            ? {
                ...offer,
                status: InStoreOfferStatus.ACTIVE,
                activeUntil: moment().add(ENV.LOCAL_OFFERS.DEFAULT_ACTIVATION_DAYS, 'days').startOf('day').toDate()
              }
            : offer
        )
      };
    case ACTION_TYPE.SET_LOCAL_OFFER_FAILED:
      return { ...state, isLocalOfferFailed: payload };
    case ACTION_TYPE.SET_HAS_CALLED_LOCAL_OFFER:
      return { ...state, hasCalledLocalOffer: payload };
    case ACTION_TYPE.SET_ROUTE_TO_ACTIVATE_LOCAL_OFFER:
      return { ...state, routeToActivateLocalOffer: payload };
    default:
      return state;
  }
};
