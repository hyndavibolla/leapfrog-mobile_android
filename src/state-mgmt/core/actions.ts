import { Props as ModalProps } from '_components/Modal';
import { Props as ToastProps } from '_components/Toast';
import { IExpirePointsBannerDataSet, TooltipKey } from '_models/general';

export enum ACTION_TYPE {
  SET_IS_DEEP_LINK_LISTENER_READY = '[core] set is deep link listener ready',
  SET_IS_APP_READY = '[core] set is app ready',
  SET_IS_CONNECTED = '[core] set is connected',
  ADD_TOAST = '[core] add toast',
  REMOVE_TOAST = '[core] remove toast',
  ADD_MODAL = '[core] add modal',
  REMOVE_MODAL = '[core] remove modal',
  SET_DEEP_LINK = '[core] set deep link',
  SET_LAST_ROUTE_KEY = '[core] set last route key',
  SET_LAST_ROUTE_PARAMS = '[core] set last route params',
  SET_ROUTE_HISTORY = '[core] set route history',
  SET_HAS_SEEN_ONBOARDING = '[core] set has seen onboarding',
  SET_VIEWED_TOOLTIP_LIST = '[core] set viewed tooltip list',
  SET_DISMISSED_TOOLTIP_LIST = '[core] set dismissed tooltip list',
  SET_HAS_SEEN_MISSION_MODAL = '[core] set has seen mission modal',
  SET_EXPIRE_POINTS_BANNER_DATA_SET = '[core] set expire points banner data set',
  SET_RZERO_USER_ID = '[core] set rzero user id',
  SET_MARKETING_ID = '[core] set marketing id',
  SET_FORCED_UPDATE_SCREEN = '[core] set force update screen',
  SET_UPDATE_TUTORIAL = '[core] set show tutorial',
  SET_TUTORIAL_AVAILABLE = '[core] set tutorial available',
  SET_TUTORIAL_FROM = '[core] set tutorial from',
  SET_EXPERIMENTS_CONFIG = '[core] set experiments config'
}

export const actions = {
  setIsDeepLinkListenerReady: (isDeepLinkListenerReady: boolean) => ({
    type: ACTION_TYPE.SET_IS_DEEP_LINK_LISTENER_READY,
    payload: { isDeepLinkListenerReady }
  }),
  setIsAppReady: (isAppReady: boolean) => ({ type: ACTION_TYPE.SET_IS_APP_READY, payload: { isAppReady } }),
  setIsConnected: (isConnected: boolean) => ({ type: ACTION_TYPE.SET_IS_CONNECTED, payload: { isConnected } }),
  addToast: (key: string, props: ToastProps) => ({ type: ACTION_TYPE.ADD_TOAST, payload: { key, props } }),
  removeToast: (key: string) => ({ type: ACTION_TYPE.REMOVE_TOAST, payload: { key } }),
  addModal: (key: string, props: ModalProps) => ({ type: ACTION_TYPE.ADD_MODAL, payload: { key, props } }),
  removeModal: (key: string) => ({ type: ACTION_TYPE.REMOVE_MODAL, payload: { key } }),
  setDeepLink: (route: string, params: Object) => ({ type: ACTION_TYPE.SET_DEEP_LINK, payload: { route, params } }),
  setLastRouteKey: (key: string) => ({ type: ACTION_TYPE.SET_LAST_ROUTE_KEY, payload: { key } }),
  setLastRouteParams: (params: object) => ({ type: ACTION_TYPE.SET_LAST_ROUTE_PARAMS, payload: { params } }),
  pushIntoRouteHistory: (route: string) => ({ type: ACTION_TYPE.SET_ROUTE_HISTORY, payload: { route } }),
  setHasSeenOnboarding: (hasSeenOnboarding: boolean) => ({ type: ACTION_TYPE.SET_HAS_SEEN_ONBOARDING, payload: { hasSeenOnboarding } }),
  setViewedTooltipList: (list: TooltipKey[]) => ({ type: ACTION_TYPE.SET_VIEWED_TOOLTIP_LIST, payload: { list } }),
  setDismissedTooltipList: (list: TooltipKey[]) => ({ type: ACTION_TYPE.SET_DISMISSED_TOOLTIP_LIST, payload: { list } }),
  setHasSeenMissionModal: (hasSeenMissionModal: boolean) => ({
    type: ACTION_TYPE.SET_HAS_SEEN_MISSION_MODAL,
    payload: { hasSeenMissionModal }
  }),
  setExpirePointsBannerDataSet: (dataSet: IExpirePointsBannerDataSet) => ({ type: ACTION_TYPE.SET_EXPIRE_POINTS_BANNER_DATA_SET, payload: { dataSet } }),
  setRzeroToken: (roToken: string) => ({ type: ACTION_TYPE.SET_RZERO_USER_ID, payload: { roToken } }),
  setMarketingId: (marketingId: string) => ({ type: ACTION_TYPE.SET_MARKETING_ID, payload: { marketingId } }),
  setForcedUpdateScreen: (showForcedUpdateScreen: boolean) => ({ type: ACTION_TYPE.SET_FORCED_UPDATE_SCREEN, payload: { showForcedUpdateScreen } }),
  showTutorial: (isTutorialVisible: boolean) => ({ type: ACTION_TYPE.SET_UPDATE_TUTORIAL, payload: { isTutorialVisible } }),
  setTutorialAvailable: (isTutorialAvailable: boolean) => ({ type: ACTION_TYPE.SET_TUTORIAL_AVAILABLE, payload: { isTutorialAvailable } }),
  setTutorialFrom: (tutorialFrom: string) => ({ type: ACTION_TYPE.SET_TUTORIAL_FROM, payload: { tutorialFrom } }),
  setExperimentsConfig: (experimentsConfig: Object) => ({ type: ACTION_TYPE.SET_EXPERIMENTS_CONFIG, payload: { experimentsConfig } })
};
