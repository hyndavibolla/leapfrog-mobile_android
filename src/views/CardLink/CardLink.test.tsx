import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react-native';

import { CardLink, Props } from './CardLink';

import { Deps, IGlobalState } from '_models/general';
import { Provider } from '_models/mission';
import { getInitialState } from '_state_mgmt/GlobalState';
import { KnownMissionSearchKey } from '_state_mgmt/mission/state';

import { getMission_1 } from '_test_utils/entities';
import { getMockDeps } from '_test_utils/getMockDeps';
import { renderWithGlobalContext } from '_test_utils/renderWithGlobalContext';
import { reducer } from '_state_mgmt/cardLink/reducer';
import { initialState as cardLinkState } from '_state_mgmt/cardLink/state';
import { actions } from '_state_mgmt/cardLink/actions';

describe('CardLink WebViewer', () => {
  let deps: Deps;
  let props: Props;
  let state: IGlobalState;

  beforeEach(() => {
    const initialStateGetter = () => {
      const initialState = getInitialState();
      initialState.mission.missionSearchMap[KnownMissionSearchKey.EXCEPTIONAL] = ['cardlink-mission'];
      initialState.mission.missionMap['cardlink-mission'] = { ...getMission_1(), provider: Provider.CARDLINK };
      return initialState;
    };

    deps = getMockDeps(initialStateGetter);
    props = {
      navigation: { goBack: jest.fn() } as any,
      route: { params: {} }
    };
    state = initialStateGetter();
  });

  it('should render', () => {
    const { toJSON } = renderWithGlobalContext(<CardLink {...props} />, deps, state);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should navigate back if url is not found', () => {
    state.mission.missionMap = {};
    renderWithGlobalContext(<CardLink {...props} />, deps, state);

    expect(props.navigation.goBack).toBeCalled();
  });

  it('should navigate back if the webView navigates to a specific finishing URL', async () => {
    deps.apiService.fetchLinkedCardsList = jest.fn();
    const { getByTestId } = renderWithGlobalContext(<CardLink {...props} />, deps, state);
    const nativeEvent = { url: 'https://www.shopyourway.com/' };
    fireEvent(getByTestId('cardlink-webview'), 'onNavigationStateChange', nativeEvent);
    await waitFor(() => {
      expect(deps.apiService.fetchLinkedCardsList).toBeCalled();
      expect(props.navigation.goBack).toBeCalled();
    });
  });

  it('should NOT navigate back if the webView navigates to a non-recognized URL', () => {
    deps.apiService.fetchLinkedCardsList = jest.fn();
    const { queryByTestId } = renderWithGlobalContext(<CardLink {...props} />, deps, state);
    const nativeEvent = { url: 'https://www.google.com/' };
    fireEvent(queryByTestId('cardlink-webview'), 'onNavigationStateChange', nativeEvent);
    expect(deps.apiService.fetchLinkedCardsList).not.toBeCalled();
    expect(props.navigation.goBack).not.toBeCalled();
  });

  it('should show a critical error when an onHttpError is fired', () => {
    const { queryByTestId } = renderWithGlobalContext(<CardLink {...props} />, deps, state);
    expect(queryByTestId('critical-error-container')).toBeFalsy();
    fireEvent(queryByTestId('cardlink-webview'), 'onHttpError', { nativeEvent: { code: 401 } });
    expect(queryByTestId('critical-error-container')).toBeTruthy();
  });

  it('should render with param shouldActivateOffer and call setRouteToActivateLocalOffer', async () => {
    deps.apiService.fetchLinkedCardsList = jest.fn();
    state.core.routeHistory = ['route1', 'route2'];
    props.route.params.shouldActivateOffer = true;
    const { getByTestId } = renderWithGlobalContext(<CardLink {...props} />, deps, state);
    const nativeEvent = { url: 'https://www.shopyourway.com/' };
    fireEvent(getByTestId('cardlink-webview'), 'onNavigationStateChange', nativeEvent);

    await waitFor(() => {
      expect(reducer({ ...cardLinkState }, actions.setRouteToActivateLocalOffer(state.core.routeHistory[1]))).toEqual({
        ...cardLinkState,
        routeToActivateLocalOffer: state.core.routeHistory[1]
      });
      expect(props.navigation.goBack).toBeCalled();
    });
  });
});
