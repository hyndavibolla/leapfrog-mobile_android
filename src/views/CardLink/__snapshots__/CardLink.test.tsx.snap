// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardLink WebViewer should render 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    accessibilityLabel="cardlink-webview"
    accessible={true}
    cacheEnabled={true}
    injectedJavaScript="true;"
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={true}
    onContentProcessDidTerminate={[Function]}
    onError={[Function]}
    onHttpError={[Function]}
    onLoad={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onNavigationStateChange={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    setSupportMultipleWindows={false}
    sharedCookiesEnabled={true}
    source={
      Object {
        "uri": "callToActionUrl_1",
      }
    }
    startInLoadingState={true}
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        undefined,
      ]
    }
    testID="cardlink-webview"
    textInteractionEnabled={true}
    thirdPartyCookiesEnabled={true}
    useSharedProcessPool={true}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F7FA",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <div>
      Spinner
    </div>
  </View>
</View>
`;
